main.rb

```diff
@@ -0,0 +1,141 @@
+# rubocop: disable Metrics/PerceivedComplexity, Metrics/CyclomaticComplexity
+module Enumerable
+  def my_each
+    return to_enum(:my_each) unless block_given?
+
+    array = to_a
+    i = 0
+    while i < array.length
+      yield(array[i])
+      i += 1
+    end
+    array
+  end
+
+  def my_each_with_index
+    return to_enum(:my_each_with_index) unless block_given?
+
+    array = to_a
+    i = 0
+    while i < array.length
+      yield(array[i], i)
+      i += 1
+    end
+    array
+  end
+
+  def my_select
+    return to_enum(:my_select) unless block_given?
+
+    new_array = []
+    my_each { |elem| new_array.push(elem) if yield(elem) == true }
+    new_array
+  end
+
+  def my_all?
+    return to_enum(:my_all?) unless block_given?
+
+    my_each { |elem| return false if yield(elem) == false }
+    true
+  end
```

Comment:
Enumerable `my_all`? when no block or argument is given returns true when none of the collection members is false or nil
     Failure/Error: `expect(true_array.my_all?).to be true_array.all?`

       expected true
            got #<Enumerator:2080> => #<Enumerator: [1, true, "hi", []]:my_all?>
     # ./enumerables_spec.rb:162:in `block (4 levels) in <top (required)>'
##
Enumerable `my_all?` when no block or argument is given returns false when one of the collection members are false or nil
     Failure/Error: `expect(false_array.my_all?).to be false_array.all?`

       expected false
            got #<Enumerator:2100> => #<Enumerator: [1, false, "hi", []]:my_all?>
     # ./enumerables_spec.rb:166:in `block (4 levels) in <top (required)>'
##
Enumerable `my_all?` when a class is passed as an argument returns true if all of the collection is a member of such class::Integer

     Failure/Error: `expect(array.my_all?(Integer)).to be array.all?(Integer)`

     ArgumentError:
       wrong number of arguments (given 1, expected 0)
     # ./enumerables_spec.rb:172:in `block (4 levels) in <top (required)>'
##
Enumerable `my_all?` when a class is passed as an argument returns true if all of the collection is a member of such class::Numeric
     Failure/Error: `expect(numbers.my_all?(Numeric)).to be numbers.all?(Numeric)`

     ArgumentError:
       wrong number of arguments (given 1, expected 0)
##
Enumerable `my_all?` when a class is passed as an argument returns false if any of the collection is not a member of such class
     Failure/Error: `expect(array.my_all?(Integer)).to be array.all?(Integer)`

     ArgumentError:
       wrong number of arguments (given 1, expected 0)
##
Enumerable `my_all?` when a Regex is passed as an argument returns true if all of the collection matches the Regex
     Failure/Error: `expect(words.my_all?(/d/)).to be words.all?(/d/)`

     ArgumentError:
       wrong number of arguments (given 1, expected 0)
##
Enumerable `my_all?` when a Regex is passed as an argument returns false if any of the collection does not match the Regex
     Failure/Error: `expect(words.my_all?(/o/)).to be words.all?(/o/)`

     ArgumentError:
       wrong number of arguments (given 1, expected 0)
##
Enumerable `my_all?` when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
      Failure/Error: `expect(array.my_all?(3)).to be array.all?(3)`

      ArgumentError:
        wrong number of arguments (given 1, expected 0)
##
Enumerable `my_all?` when a pattern other than Regex or a Class is given returns false if any of the collection does not match the pattern
      Failure/Error: `expect(array.my_all?(3)).to be array.all?(3)`

      ArgumentError:
        wrong number of arguments (given 1, expected 0)
##